src/statsdproto.rs

```diff
@@ -0,0 +1,385 @@
+use bytes::BufMut;
+use bytes::Bytes;
+use memchr::{memchr, memchr2};
+use thiserror::Error;
+
+use std::{
+    convert::{TryFrom, TryInto},
+    vec,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Type {
+    Counter,
+    Timer,
+    Gauge,
+    DirectGauge,
+    Set,
+}
+
+impl TryFrom<&[u8]> for Type {
+    type Error = ParseError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() < 1 && value.len() > 2 {
+            return Err(ParseError::InvalidType);
+        }
+        match value[0] {
+            b'c' => Ok(Type::Counter),
+            b'm' => Ok(Type::Timer),
+            b'g' => Ok(Type::Gauge),
+            b'G' => Ok(Type::DirectGauge),
+            b's' => Ok(Type::Set),
+            _ => Err(ParseError::InvalidType),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("invalid parsed value")]
+    InvalidValue,
+    #[error("invalid sample rate")]
+    InvalidSampleRate,
+    #[error("invalid type")]
+    InvalidType,
+    #[error("invalid tag")]
+    InvalidTag,
+}
+
+/// A Tag is a key:value pair of metric tags
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub struct Tag {
```

Comment:
Do you think there's value in pushing the tag parsing logic down to the PDU layer?

I think in order to keep the allocation-free property you'd have to express it as an iterator, yielding consecutive byte slice references, but that might make for a meaningful expressiveness improvement in the lower level structure while turning the `parse_tags` method into effectively a collect call.

Reply:
Yeah, I think there is value here. Its grafted on the Owned type since I wrote that second, but from a byte-range emitting parser it may be useful to factor this down into the PDU phase. Let me revise that a bit.